oak/server/rust/oak_runtime/src/runtime.rs

```diff
@@ -0,0 +1,133 @@
+
+use oak_abi::OakStatus;
+
+use crate::channel;
+use crate::channel::{ChannelReader, ChannelWriter};
+use crate::node::{load_wasm, NodeConfiguration};
+use crate::platform::*;
+use crate::proto;
+use crate::proto::manager::NodeConfiguration_oneof_config_type;
+
+/// Runtime structure for configuring and running a set of Oak nodes.
+pub struct Runtime {
+    configurations: HashMap<String, NodeConfiguration>,
+    terminating: AtomicBool,
+    node_threads: Mutex<Vec<std::thread::JoinHandle<()>>>,
+}
+
+impl Runtime {
+    /// Configure and run the protobuf specified `ApplicationConfiguration`. After creating a
+    /// `Runtime` calling `stop` will send termination signals to nodes and wait for them to
+    /// terminate.
+    pub fn configure_and_run(
+        app_config: proto::manager::ApplicationConfiguration,
+    ) -> Result<(RuntimeRef, ChannelWriter), OakStatus> {
+        let mut runtime = Runtime {
+            configurations: HashMap::new(),
+            terminating: AtomicBool::new(false),
+            node_threads: Mutex::new(Vec::new()),
+        };
+
+        for config in app_config.get_node_configs() {
+            runtime.configurations.insert(
+                config.name.clone(),
+                match &config.config_type {
+                    None => {
+                        error!("Node config {} with no type", config.name);
+                        return Err(OakStatus::ERR_INVALID_ARGS);
+                    }
+                    Some(NodeConfiguration_oneof_config_type::log_config(_)) => {
+                        NodeConfiguration::LogNode
+                    }
+                    Some(NodeConfiguration_oneof_config_type::wasm_config(
+                        proto::manager::WebAssemblyConfiguration { module_bytes, .. },
+                    )) => load_wasm(&module_bytes),
+                    Some(_) => {
+                        error!("Pseudo-node not implemented!");
+                        return Err(OakStatus::ERR_INVALID_ARGS);
+                    }
+                },
+            );
+        }
+
+        let runtime = RuntimeRef(Arc::new(runtime));
+        let (chan_writer, chan_reader) = channel::new();
+
+        runtime.node_create(
+            &app_config.initial_node_config_name,
+            &app_config.initial_entrypoint_name,
+            chan_reader,
+        )?;
+
+        Ok((runtime, chan_writer))
+    }
+
+    /// Thread safe method for determining if the `Runtime` is terminating.
+    pub fn is_terminating(&self) -> bool {
+        self.terminating.load(SeqCst)
+    }
+
+    /// Thread safe method for signaling termination to a `Runtime` and waiting for it's node
+    /// threads to terminate.
+    pub fn stop(&self) {
+        let mut node_threads = self.node_threads.lock().unwrap();
+
+        self.terminating.store(true, SeqCst);
+
+        let mut handles = vec![];
+        std::mem::swap(&mut handles, &mut node_threads);
+        let _ = handles.into_iter().map(|h| h.join());
```

Comment:
Does this have any effect at all? I think the compiler would just optimise it away, it should probably be an explicit `for` loop.

Reply:
Good catch, this should of at least had a `collect` which does fix the issue- let me know if you think a `for` loop would be better/more readable than this though

Reply:
Using `collect` to force side effects is not very idiomatic, so I would strongly suggest an explicit `for` loop. Note that I also suspect even your later code is still optimized away, since the return value of the collected result is still unused.

Reply:
Done, used a for loop for clarity